bin/test/i18n/utils/test_sync_up_base.rb

```diff
@@ -0,0 +1,219 @@
+      end
+
+      context 'when new value is pushed' do
+        let(:described_class) do
+          Class.new(I18n::Utils::SyncUpBase) do
+            config.source_paths << 'new_source_path'
+          end
+        end
+
+        it 'returns array with new value' do
+          assert_equal ['new_source_path'], source_paths
+        end
+      end
+    end
+
+    describe '#ignore_paths' do
+      let(:ignore_paths) {config.ignore_paths}
+
+      it 'returns empty array' do
+        assert_equal [], ignore_paths
+      end
+
+      context 'when new value is pushed' do
+        let(:described_class) do
+          Class.new(I18n::Utils::SyncUpBase) do
+            config.ignore_paths << 'new_ignored_path'
+          end
+        end
+
+        it 'returns array with new value' do
+          assert_equal ['new_ignored_path'], ignore_paths
+        end
+      end
+    end
+  end
+
+  describe '.perform' do
+    let(:perform) {described_class.perform}
+
+    it 'creates new instance and calls #perform' do
+      described_class.any_instance.expects(:perform).once
+      perform
+    end
+
+    it 'calls report_runtime metrics with class name' do
+      I18n::Metrics.expects(:report_runtime).with(described_class.name, 'sync-up').once
+      perform
+    end
+
+    context 'when class name contains i18n resource name' do
+      module I18n
+        module Resources
+          module ResourceParent
+            module ResourceChild
+              class SyncUp < I18n::Utils::SyncUpBase
+              end
+            end
+          end
+        end
+      end
+
+      let(:described_class) {I18n::Resources::ResourceParent::ResourceChild::SyncUp}
+
+      it 'calls report_runtime metrics with i18n resource name' do
+        I18n::Metrics.expects(:report_runtime).with('ResourceParent::ResourceChild', 'sync-up').once
+        perform
+      end
+    end
+  end
+
+  describe '#perform' do
+    let(:perform) {described_instance.send(:perform)}
+
+    let(:expected_crowdin_client) {stub}
+    let(:expected_source_file_path) {'expected_source_file_path'}
+
+    before do
+      described_instance.stubs(:crowdin_client).returns(expected_crowdin_client)
+      described_instance.stubs(:source_files).returns([expected_source_file_path])
+    end
+
+    it 'uploads source file' do
+      execution_sequence = sequence('execution')
+
+      described_instance.expects(:crowdin_client).in_sequence(execution_sequence).returns(expected_crowdin_client)
+      expected_crowdin_client.expects(:upload_source_file).with(expected_source_file_path).in_sequence(execution_sequence)
+
+      perform
+    end
+  end
+
+  describe '#config' do
+    let(:config) {described_instance.send(:config)}
+
+    let(:expected_class_config) {'expected_class_config'}
+
+    before do
+      described_class.expects(:config).returns(expected_class_config)
+    end
+
+    it 'returns class config' do
+      assert_equal expected_class_config, config
+    end
+  end
```

Comment:
We use stubbing to abstract from the business logic of the whole class and focus on (unit) testing the code of this particular method.

In this case, we check that the instance method `#config` works as an alias and simply returns the class method `.config` value and nothing else.

And in the test for the class method `.config` we check if the value is correct.

Reply:
For example, if we use the real class method `.config` for testing instead of the stub,
the test will fail when something is wrong with `.config` (e.g., if it raises an exception).

```ruby  
describe '#config' do
  let(:config) {described_instance.send(:config)}

  it 'returns class config' do
    assert_equal described_class.config, config
  end
end
```

This would be okay in integration testing but not in unit testing because the method code technically still works correctly - it calls the class method `.config` and returns its value.

So, this would prevent us from quickly identifying the real trouble spot - `.config`.

Reply:
My concern is that we _aren't_ unit testing the code of this particular method, because we stub out the method before testing. Notably, if I make this change, the test still passes:

```diff
diff --git a/bin/i18n/utils/sync_up_base.rb b/bin/i18n/utils/sync_up_base.rb
index 09460f6cc1f..6cfd0237104 100644
--- a/bin/i18n/utils/sync_up_base.rb
+++ b/bin/i18n/utils/sync_up_base.rb
@@ -10,11 +10,7 @@ module I18n
       MAIN_CROWDIN_PROJECT = 'codeorg'.freeze
 
       def self.config
-        @config ||= Config.new(
-          crowdin_project: MAIN_CROWDIN_PROJECT,
-          source_paths: [],
-          ignore_paths: [],
-        )
+        raise NotImplementedError
       end
 
       def self.perform
```

Ironically, if I do so then several *other* tests start failing, because those tests are indeed dependent on the config method working. It seems to me that this test is generating more noise than signal, as it were; I'd recommend removing it.